docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+  locality guarantees
+
+Terminology
+===========
+
+Since different projects have used different words to describe various
+concepts, here is a small glossary to help disambiguate.
+
+* **Array** or **Vector**: a sequence of values with known length all
+  having the same type. These terms are used interchangeably in
+  different Arrow implementations, but we use "array" in this
+  document.
+* **Slot**: a single logical value in an array of some particular data type
+* **Buffer** or **Contiguous memory region**: a sequential virtual
+  address space with a given length. Any byte can be reached via a
+  single pointer offset less than the region's length.
+* **Physical Layout type**: The underlying memory layout for an array
+  without taking into account any value semantics. For example, a
+  32-bit signed integer array and 32-bit floating point array have the
+  same layout type.
+* **Logical type**: An application-facing semantic value type that is
+  implemented using some physical layout type. For example, Decimal
+  values are stored as 16 bytes in a fixed-size binary
+  layout. Similarly, strings can be stored as ``List<1-byte>``. A
+  timestamp may be stored as 64-bit fixed-size layout.
+* **Nested** or **parametric type**: a data type whose full structure
+  depends on one or more other child types. Two fully-specified nested
+  types are equal if and only if their child types are equal. For
+  example, ``List<U>`` is distinct from ``List<V>`` iff U and V are
+  different types.
+* **Parent** and **child arrays**: names to express relationships
+  between physical value arrays in a nested type structure. For
+  example, a ``List<T>``-type parent array has a T-type array as its
+  child (see more on lists below).
+* **Leaf node** or **leaf**: A primitive value array that may or may
+  not be a child array of some array with a nested type.
+
+Physical Memory Layout
+======================
+
+Arrays are defined by a few pieces of metadata and data:
+
+* A logical data type
+* A sequence of buffers
+* A length as a 64-bit signed integer. Implementations are permitted
+  to be limited to 32-bit lengths, see more on this below
+* A null count as a 64-bit signed integer
+* An optional **dictionary**, for dictionary-encoded arrays
+
+Nested arrays additionally have a sequence of one or more sets of
+these items, called the **child arrays**.
+
+Each logical data type has a well-defined physical layout. Here are
+the different physical layouts defined by Arrow:
+
+* **Primitive (fixed-size)**: a sequence of values each having the
+  same byte or bit width
+* **Variable-size Binary**: a sequence of values each having a variable
+  byte length. Two variants of this layout are supported using 32-bit
+  and 64-bit length encoding.
+* **Fixed-size List**: a nested layout each each value has the same
+  number of elements taken from a child data type.
+* **Variable-size List**: a nested layout type where each value is a
```

Comment:
Seems like sometimes we're using "layout" and sometimes "layout type". It would be nice to be consistent. Varying terminology often impedes understanding.